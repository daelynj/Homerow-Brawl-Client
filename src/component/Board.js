import React, { useState } from 'react';
import TextBox from './textbox/TextBox';
import TypingBox from './typingbox/TypingBox';
import Timer from './Timer'
import WordsPerMinute from './WordsPerMinute';

function Board(props) {
  const [words] = useState(props.text.split(" "));
  const [currentWord, setCurrentWord] = useState(0);
  const [currentInput, setCurrentInput] = useState("");
  const [finishTime, setFinishTime] = useState(0);
  const [start, setStart] = useState(false);

  //this if statement only allows the text box to render while we have words
  //to type so that we don't slice non-existent text (this results in a crash)
  function renderTextBox() {
    if (currentWord < words.length) {
      let textToCompare = words[currentWord].slice(0, currentInput.length);
      return (
        <TextBox
          words = {words}
          input = {currentInput}
          textToCompare = {textToCompare}
          currentWord = {currentWord}
        />
      )
    }
  }

  function renderTypingBox() {
    if (currentWord < words.length) {
      return (
        <TypingBox
          value = {currentInput}
          onChange = {setCurrentInput}
          onWordComplete = {checkWord}
          start = {setStart}
        />
      )
    }
  }

  function checkWord() {
    if (currentInput === words[currentWord]) {
      setCurrentInput('');
      setCurrentWord(currentWord+1);
      return true;
    }
  }

  //only start the timer when a key is hit
  function renderTimer() {
    if (start) {
      if (currentWord < words.length) {
        return (
          <Timer
            time = {0}
            finishTime = {setFinishTime}
          />
        )
      }
      else {
        return (
          <WordsPerMinute
            wordsLength = {words.length}
            finishTime = {finishTime}
          />
        )
      }
    }
  }

  return (
    <div className="boxes">
      <div className="text-box">
        {renderTextBox()}
      </div>
      <div className='typing-box'>
        {renderTypingBox()}
      </div>
      <div className="timer">
        {renderTimer()}
      </div>
    </div>
  )
}

export default Board;
