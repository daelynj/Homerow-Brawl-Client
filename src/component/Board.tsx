import * as React from "react";
import { Race } from "./race/Race";
import { Text } from "./text/Text";
import { TypingBox } from "./typingbox/TypingBox";
import { Timer } from "./Timer/Timer";
import { GameStats } from "./gamestats/GameStats";

interface Props {
  words: String[];
  currentWordSubstring: any;
  currentInput: string;
  setCurrentInput: (newInput: string) => void;
  currentWordIndex: number;
  setCurrentWordIndex: (newIndex: number) => void;
  start: boolean;
  setStart: (newStart: boolean) => void;
  endGame: () => boolean;
  finishTime: number;
  setFinishTime: (newTime: number) => void;
  checkLetter: () => void;
  incorrectLetters: number;
  updatePosition: (newPosition: any) => void;
  raceState: any;
  ID: number;
}

export const Board = (props: Props) => {
  const renderRace = () => (
    <Race
      ID={props.ID}
      raceState={props.raceState}
      updatePosition={props.updatePosition}
      wordsLength={props.words.length}
      currentWordIndex={props.currentWordIndex}
    />
  );

  const renderText = () => (
    <Text
      words={props.words}
      input={props.currentInput}
      currentWordSubstring={props.currentWordSubstring}
      currentWordIndex={props.currentWordIndex}
    />
  );

  const renderTypingBox = () => (
    <TypingBox
      value={props.currentInput}
      onChange={props.setCurrentInput}
      setStart={props.setStart}
      word={props.words[props.currentWordIndex]}
      setCurrentWordIndex={props.setCurrentWordIndex}
      currentWordIndex={props.currentWordIndex}
      checkLetter={props.checkLetter}
    />
  );

  const renderTimer = () => <Timer setFinishTime={props.setFinishTime} />;

  const renderGameStats = () => (
    <GameStats
      words={props.words}
      finishTime={props.finishTime}
      incorrectLetters={props.incorrectLetters}
    />
  );

  const shouldRenderTextAndInput = () => (!props.endGame() ? true : false);

  const shouldRenderTimer = () =>
    !props.endGame() && props.start ? true : false;

  const shouldRenderGameStats = () => (props.endGame() ? true : false);

  return (
    <div className="board">
      <div className="race">{renderRace()}</div>
      <div className="text">{shouldRenderTextAndInput() && renderText()}</div>
      <div className="typing-box">
        {shouldRenderTextAndInput() && renderTypingBox()}
      </div>
      <div className="timer">{shouldRenderTimer() && renderTimer()}</div>
      <div className="gamestats">
        {shouldRenderGameStats() && renderGameStats()}
      </div>
    </div>
  );
};
