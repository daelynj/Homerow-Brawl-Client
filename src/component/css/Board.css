.input {
  border: 5px solid teal;
  margin-left: auto;
  margin-right: auto;
  width: 55%;
}
