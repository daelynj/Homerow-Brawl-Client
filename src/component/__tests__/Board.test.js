import React from "react";
import { Board } from "../Board";
import { mount, configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

configure({ adapter: new Adapter() });

describe("Board", () => {
  const setStart = jest.fn();
  const setCurrentInput = jest.fn();
  const setCurrentWordIndex = jest.fn();
  const setFinishTime = jest.fn();
  const endGame = jest.fn();
  const checkLetter = jest.fn();

  const buildProps = (newProps = {}) => ({
    words: ["this", "is", "text"],
    start: false,
    setStart,
    currentWordSubstring: "",
    currentInput: "",
    setCurrentInput,
    currentWordIndex: 0,
    setCurrentWordIndex,
    endGame,
    finishTime: 0,
    setFinishTime,
    checkLetter,
    incorrectLetters: 0,
    ...newProps
  });

  describe("the game before it's running", () => {
    it("renders the expected components", () => {
      const BoardComponent = mount(<Board {...buildProps()} />);

      expect(BoardComponent.find("Text").exists()).toBe(true);
      expect(BoardComponent.find("TypingBox").exists()).toBe(true);
      expect(BoardComponent.find("Race").exists()).toBe(true);

      expect(BoardComponent.find("Timer").exists()).toBe(false);
      expect(BoardComponent.find("GameStats").exists()).toBe(false);
    });
  });

  describe("the game once it's running", () => {
    it("renders the expected components", () => {
      const BoardComponent = mount(<Board {...buildProps({ start: true })} />);

      expect(BoardComponent.find("Text").exists()).toBe(true);
      expect(BoardComponent.find("TypingBox").exists()).toBe(true);
      expect(BoardComponent.find("Timer").exists()).toBe(true);
      expect(BoardComponent.find("Race").exists()).toBe(true);

      expect(BoardComponent.find("GameStats").exists()).toBe(false);
    });
  });

  describe("the game when it is finished", () => {
    it("renders the expected components", () => {
      const endGame = jest.fn().mockReturnValue(true);
      const BoardComponent = mount(
        <Board {...buildProps({ start: true, endGame: endGame })} />
      );

      expect(BoardComponent.find("GameStats").exists()).toBe(true);
      expect(BoardComponent.find("Race").exists()).toBe(true);

      expect(BoardComponent.find("Text").exists()).toBe(false);
      expect(BoardComponent.find("TypingBox").exists()).toBe(false);
      expect(BoardComponent.find("Timer").exists()).toBe(false);
    });
  });
});
