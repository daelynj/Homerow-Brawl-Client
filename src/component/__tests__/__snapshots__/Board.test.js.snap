// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board the game before it's running renders the expected components 1`] = `
<div
  className="board"
>
  <div
    className="text"
  >
    <div
      className="Words"
    >
      <span
        className="Word"
      >
        <span
          className="black"
        >
          t
        </span>
        <span
          className="black"
        >
          h
        </span>
        <span
          className="black"
        >
          i
        </span>
        <span
          className="black"
        >
          s
        </span>
      </span>
      <span
        className="Word"
      >
        <span
          className="black"
        >
          i
        </span>
        <span
          className="black"
        >
          s
        </span>
      </span>
      <span
        className="Word"
      >
        <span
          className="black"
        >
          t
        </span>
        <span
          className="black"
        >
          e
        </span>
        <span
          className="black"
        >
          x
        </span>
        <span
          className="black"
        >
          t
        </span>
      </span>
    </div>
  </div>
  <div
    className="typing-box"
  >
    <div
      className="TypingBox"
    >
      <input
        onChange={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    className="timer"
  />
  <div
    className="wordsperminute"
  />
</div>
`;

exports[`Board the game once it's running renders the expected components 1`] = `
<div
  className="board"
>
  <div
    className="text"
  >
    <div
      className="Words"
    >
      <span
        className="Word"
      >
        <span
          className="black"
        >
          t
        </span>
        <span
          className="black"
        >
          h
        </span>
        <span
          className="black"
        >
          i
        </span>
        <span
          className="black"
        >
          s
        </span>
      </span>
      <span
        className="Word"
      >
        <span
          className="black"
        >
          i
        </span>
        <span
          className="black"
        >
          s
        </span>
      </span>
      <span
        className="Word"
      >
        <span
          className="black"
        >
          t
        </span>
        <span
          className="black"
        >
          e
        </span>
        <span
          className="black"
        >
          x
        </span>
        <span
          className="black"
        >
          t
        </span>
      </span>
    </div>
  </div>
  <div
    className="typing-box"
  >
    <div
      className="TypingBox"
    >
      <input
        onChange={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    className="timer"
  >
    <div>
      Timer: 
      0
    </div>
  </div>
  <div
    className="wordsperminute"
  />
</div>
`;

exports[`Board the game when it is finished renders the expected components 1`] = `
<div
  className="board"
>
  <div
    className="text"
  />
  <div
    className="typing-box"
  />
  <div
    className="timer"
  />
  <div
    className="wordsperminute"
  >
    <div
      className="wpm"
    >
      WPM: Infinity
    </div>
  </div>
</div>
`;
